gelic-idealab/impress
# World Pulling Fixes

# World Pulling Fixes

## Scale, Rotate, Position

You can now grab the world and expect those two points of the world to stay positioned where your hands are. As a result, it does feel more like pulling the world. (Remember, though, you are effectively transforming yourself in the inverse way.)

Additionally, this is fully compatible with the snap turns feature; the bug where snap-turning before world pulling would result in broken rotation is now repaired.

### Type of Change


### How Has This Been Tested?


Steps to test: 

1. Use Unity with a Link, Rift, or Rift S.
1. Enter Play mode. 
1. Squeeze both triggers to begin world pulling. 
1. Move your hands further apart or close together to scale.
1. Rotate your hands like spinning a plate by the edges to rotate.
1. Move your hands to move the world.

Expected result: The world appears to move as if you are grabbing it.

## Increased Scale Range, Repaired Ruler

The scale of the player now goes from 0.1m to 10m. This is configurable in the inspector. If the player's starts outside the range specified, it will default to 0.1m min and 10m max. Calculating the texture position on the animal ruler (taken from Google Tilt Brush) has now been refactored for clarity.

### Type of Change


### How Has This Been Tested?

Steps to test: 

1. Enter Play mode. 
2. Perform world pulling scaling.
3. Try to scale yourself large.
4. Try to scale yourself small. 

Expected result: The scale is clamped to a minimum and maximum, and the dinosaur is shown at the largest scale and the squirrel is shown at the smallest scale.


## Debug Axes

The position and rotation (and sometimes scale) of both utility objects created by the world pulling code and elements of the XR rig can be observed with the debug axes feature. To use it, access the World Pulling gameObject in the scene, and check the "Show Debug Axes" checkbox.

### Type of Change


### How Has This Been Tested?

See above.

Expected result: Some debug axes show the initial positions of the playspace, pivot point local to the playspace, and global pivot point. The remaining debug axes show the current transforms of the playspace, pivot point local to the playspace, and hands. 

____

## Test Configuration

No browser yet

Oculus Quest with Link

Unity Editor

Windows 10

# Checklist:


Modified from this article:
Phillip Johnston, “A GitHub Pull Request Template for Your Projects - Embedded Artistry,” Embedded Artistry, Aug. 04, 2017. https://embeddedartistry.com/blog/2017/08/04/a-github-pull-request-template-for-your-projects/ (accessed Jul. 22, 2021).
